### DISCUSSION TITLE
Receiving special events through WebSocket

### DISCUSSION THREAD

**Discussion by zaalgol** [getgauge/taiko#2035] [Development]
Hi,
My code browses the web through a proxy, while the proxy returns an application that you browse with and receives messages through a  WebSocket.

I succeed in tests to browsing this way, but do not get messages from the WebSocket.
I am waiting for a special event from the server called 'TestEvent' in the following code, and it is not received.
I tried inside the goto, and also inside waitFor

code:
 
```
 await openBrowser({
            args: [
                '--disable-gpu',
                '--disable-dev-shm-usage',
                '--disable-setuid-sandbox',
                '--no-first-run',
                '--no-sandbox',
                '--no-zygote',
                '--proxy-server=http://localhost:8080'
            ],
            headless: true,
        })

await goto(url,
            {
                navigationTimeout:10000,
               // waitForEvents:['TestEvent'], // ERROR!
                headers: { 'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome' }
            });
           // await waitFor('TestEvent'); // ERROR!
```

post Scriptum: In Puppeteer it works with:

```
await page.exposeFunction('onCustomEvent', ({ detail }) => {
    // ...
    }).catch((error) => {
        console.log("exposeFunction error:", error);
    });
 
    await page.evaluateOnNewDocument(() => {
        window.addEventListener('TestEvent', ({ detail }) => {
            window.onCustomEvent({ detail });
        });
    }).catch((error) => {
        console.log("evaluateOnNewDocument error:", error);
    });
```
 
Is there anything similar or equivalent in Taiko?

**Comment by zabil** [comment]
Have you tried using [`evaluate`](https://docs.taiko.dev/running_scripts_inside_the_browser/#perform-actions-on-the-page)?

**Reply by zaalgol** [reply]
Yes.
```
await evaluate(() => {
            window.addEventListener('TestEvent', ({ detail }) => {
                // ...
            });
        });
```
But still don't get web socket messages

**Reply by NivedhaSenthil** [reply]
anything that is done within evaluate will be within the browser context it has to be explicitly sent back to taiko if we want to access them.. can you please tell us how are you trying to handle the listener ?

**Reply by zaalgol** [reply]
```
let incommingMsg ="":
//...
await evaluate(() => {
            window.addEventListener('TestEvent', ({ detail }) => {
               incommingMsg = details;
        });
});
```

**Reply by NivedhaSenthil** [reply]
the incommingMsg is in taiko's context and it wont be updated via evaluate , you will have to handle it in the browser context itself.. try doing a console log in the listener and check if you could see it in browser's console

**Reply by zaalgol** [reply]
> the incommingMsg is in taiko's context and it wont be updated via evaluate , you will have to handle it in the browser context itself.. try doing a console log in the listener and check if you could see it in browser's console

I was able to log the incoming message like this:

```
let incommingMsg ="":
//...
await evaluate(() => {
            window.addEventListener('TestEvent', ({ detail }) => {
                console.log(details)
                incommingMsg = details; // incommingMsg  value doesn't change
        });
});
```

The problem is that incommingMsg variable doesn't change.

If and how can I transfer data between the browser's context and Taiko's context? Or, if I can call a function that resides in Taiko's context from the browser's context?
 
 Thanks

**Reply by zabil** [reply]
There's a section in the docs about passing data between the browser and Taiko

https://docs.taiko.dev/running_scripts_inside_the_browser/#passing-data

In your case however in your case it's a bit tricky as the message is set in an event. So you might have to make two calls.

```
await evaluate(() => {
            window.addEventListener('TestEvent', ({ detail }) => {
                console.log(details)
                window.incommingMsg = details; // incommingMsg  value doesn't change
        });
});

await evaluate(() => {
                return window.incommingMsg;
});
```

As this depends on `TestEvent ` event being raise you might need to wait for variable.

**Reply by elig1984** [reply]
Hi, 
I'm working with zaalgol and am also interested in this issue.
In the code you suggested in your last reply, the line:
console.log 
is reached multiple times > every time the event 'TestEvent' is raised.
However, the 2nd evaluate will happen only once.
What we're trying to do is to be able to evaluate the event data (the "details" variable) in the taiko context each and every time the TestEvent is raised.
To my understanding, in order for that to happen, I need to be able from within the event handler (where console.log is happening) to call a taiko function. Is there a way to do that? That is exactly what exposeFunction in Puppateer enables me to do. It exposes a server side function to the browser context.
Thanks

**Reply by zabil** [reply]
I don't think there's an API on Taiko for that.
However you can interact with [CDP](https://chromedevtools.github.io/devtools-protocol/tot/DOMDebugger/#type-EventListener) via Taiko's [client](https://docs.taiko.dev/api/client/) API. Maybe there's some way you can write a custom function to handle this.
